dhis-2/dhis-services/dhis-service-program-rule/src/main/java/org/hisp/dhis/programrule/engine/DefaultProgramRuleEntityMapperService.java

```diff
@@ -472,29 +256,41 @@ public RuleEvent toMappedRuleEvent(Event eventToEvaluate) {
     String orgUnit = getOrgUnit(eventToEvaluate);
     String orgUnitCode = getOrgUnitCode(eventToEvaluate);
 
-    return RuleEvent.create(
+    return new RuleEvent(
         eventToEvaluate.getUid(),
         eventToEvaluate.getProgramStage().getUid(),
+        eventToEvaluate.getProgramStage().getName(),
         RuleEvent.Status.valueOf(eventToEvaluate.getStatus().toString()),
-        ObjectUtils.defaultIfNull(
-            eventToEvaluate.getOccurredDate(), eventToEvaluate.getScheduledDate()),
-        eventToEvaluate.getScheduledDate(),
+        eventToEvaluate.getOccurredDate() != null
+            ? Instant.Companion.fromEpochMilliseconds(eventToEvaluate.getOccurredDate().getTime())
+            : Instant.Companion.fromEpochMilliseconds(eventToEvaluate.getScheduledDate().getTime()),
```

Comment:
`scheduledDate` is user-provided input. Some validation around `scheduledDate` may be useful. 
In case we have invalid input we can skip the event with a message in the logs.

Reply:
The assumption here is that the payload (events and enrollments) is already validated.
We are running rule-engine after validation in tracker.
If an event is invalid, it will not be part of the payload used to run rule-engine